packages/jwt-middleware/index.js

```diff
@@ -0,0 +1,88 @@
+'use strict';
+
+const Assert = require('assert');
+const ExpressJwt = require('express-jwt');
+const jws = require('jws');
+const JwksRsa = require('jwks-rsa');
+
+const WT_AUTH_EXEC = 'wt-authorize-execution';
+
+module.exports = () => {
+
+    return function(req, res, next) {
+
+        const ctx = req.webtaskContext;
+    
+        const iss = ctx.meta['iss'];
+        const aud = ctx.meta['aud'];
+
+        // if wt-authorize-execution is set to true (non-zero) then proceed with authn and authz
+
+        if (ctx.meta[WT_AUTH_EXEC] && ctx.meta[WT_AUTH_EXEC] !== "0"){
+
+            if (req.headers['authorization']) {
+
+                const match = (req.headers['authorization'] || '')
+                .trim()
+                .match(/^bearer (.+)$/i);
+
+                let jwt;
+
+                try {
+                    jwt = jws.decode(match[1]);
+                    if (!jwt) throw new Error();
+                } catch(e){
+                    const error = new Error('Invalid authorization header');
+                    error.statusCode = 401;
+                    return next(error);
+                }
+
+                // verify claims
+
+                if (
+                    jwt.payload.scope.indexOf("wt:admin") === -1 
+                    && jwt.payload.scope.indexOf(`wt:owner:${req.x_wt.container}`) === -1
+                    && jwt.payload.scope.indexOf(ctx.meta['wt-execution-scope']) === -1
+                ) {
+                    const error = new Error('Missing required scopes');
+                    error.statusCode = 401;
+                    return next(error);
+                }
+
+            } else {
+                const error = new Error('Missing authorization header');
+                error.statusCode = 401;
+                return next(error);
+            }
```

Comment:
Normally the authorization check is only done after authentication checks have completed (i.e. signature validation). Otherwise unauthenticated users may test the authorization rules.

Reply:
This is one part I want to revisit.  The Authentication is using `express_jwt` and if it passes it will return `next()` and skip rest of my Authorization code.  I tried to flip the order using `async.series` without any luck.